airflow/utils/timeout.py

```diff
@@ -31,20 +34,34 @@ def __init__(self, seconds=1, error_message='Timeout'):
         self.seconds = seconds
         self.error_message = error_message + ', PID: ' + str(os.getpid())
 
-    def handle_timeout(self, signum, frame):  # pylint: disable=unused-argument
+    def handle_timeout(self, *args):  # pylint: disable=unused-argument
         """Logs information and raises AirflowTaskTimeout."""
         self.log.error("Process timed out, PID: %s", str(os.getpid()))
         raise AirflowTaskTimeout(self.error_message)
 
     def __enter__(self):
         try:
-            signal.signal(signal.SIGALRM, self.handle_timeout)
-            signal.setitimer(signal.ITIMER_REAL, self.seconds)
+            if IS_WINDOWS:
+                if hasattr(self, TIMER_THREAD_ATTR) and getattr(self, TIMER_THREAD_ATTR) is not None:
+                    getattr(self, TIMER_THREAD_ATTR).cancel()
+                timer = Timer(self.seconds, self.handle_timeout)
+                setattr(self, TIMER_THREAD_ATTR, timer)
+                timer.start()
```

Comment:
The exception message below (about hte current context) doesn’t apply to the thread-based implementation, so we should move the `if IS_WINDOWS:` block out of `try:`.

Reply:
BTW why do we need to use `has|get|set_attr` here? Why not put them on the instance instead, they would just be unused on POSIX. Or we can do something like

```python
_timeout = ContextManager[None]

class _timeout_windows(_timeout):
   ...  # Implementation for Windows.

class _timeout_posix(_timeout):
   ...  # Implementation for POSIX.

if IS_WINDOWS:
    timeout: Type[_timeout] = _timeout_windows
else:
    timeout = _timeout_posix
```

Reply:
I really like this approach, much cleaner than the current one. :)